### ISSUE TITLE
Category filters on dashboard not working for postgres in v0.28.0 TEXT data type

### ISSUE THREAD

**Issue by cmann50** [metabase/metabase#6914]
### Bugs
If you're filing an issue about a bug please include as much information
as you can including the following.

- Your browser and the version: Chrome
- Your operating system: OSX High Siera
- Your databases:  Postgres
- Metabase version:  0.28.0
- Metabase hosting environment: Jar deploy to Ubuntu 16
- Metabase internal database: H2 (default)

- *Repeatable steps to reproduce the issue*
 
I upgraded to 0.28.0 today when it came out and one of my category filter drop downs wasn't working.  It kept switching to a free form text.  I added all my columns to the dashboard and the drop down was working fine for each column.   I then realized the broken column was the only once of type TEXT in postgres.

I fixed the problem by:
- Changing the column to type varchar(255) instead of TEXT in postgres
- In metabase admin panel, go to database > choose your DB > and sync it and rescan field values
- Go to Admin > data model > select the table on the left, then click the little gear icon in the upper right, and choose _discard cached field values_, and _rescan the table_

After those steps the category filter on the dashboard switched over to a drop down and showed the data in the table.

It's possible it was broken on 0.27.0 and I has switched that field to TEXT type recently and just didn't notice it since I upgraded to 0.28.0 when it came out today.

Hopefully this workaround will save someone some time.

![image](https://user-images.githubusercontent.com/2318067/36009232-4b486f7e-0d19-11e8-8521-c1d0dd6ea886.png)

**Comment by salsakran** [comment]
How many distinct values are/were in that field?

**Comment by cmann50** [comment]
I should have updated this ticket earlier.  I thought this solved my problem, but it turned out I had too many rows in the table ( I think around 468 rows - I saw on another ticket the limit was 300) and so the category drop down was changing.  The reason it appeared to be fixed is because I was loading in data at the time.  

I ended up splitting the table into separate tables using views as a workaround for now.  This gets the rows in the table low enough for the category drop down to start working again.  

It's not a good solution because I need different tables (views) per team and different dashboards.   I couldn't find a way using segments or other metabase features to split the data (using segments etc).  

I also pondered rotating the rows into columns using postgres crosstab function but decided it wouldn't work very well.  Let me know if you know of a better idea.

**Comment by salsakran** [comment]
we're changing the way the category/autocomplete mix works in #4241 regardless, so in 0.29 situations like this shouldn't happen

**Comment by cmann50** [comment]
Awesome!  Looking forward to trying it out.

**Comment by ArghyaSarkar** [comment]
Browser: Chrome Version 64.0.3282.186
Operating system : OSX Siera 10.12.6
Database: Postgres
Metabase version: 0.28.1
Metabase hosting environment: 
Metabase internal database: H2 (default)

Repeatable steps to reproduce the issue

I have a table `abc_delivery_records` which has 5 columns.

```
id - int, primary key
delivery_date - date
station_code - varchar, null: false
delivery_mode - varchar, null: true
num_deliveries - int
```

I wrote a simple query like below.

```
SELECT station_code, COUNT(num_deliveries)
FROM abc_delivery_records
WHERE TRUE
  [[AND {{station_code}}]]
  [[AND {{delivery_mode}}]]
GROUP BY 1
```

The field filter widget was not visible for any of the filters applied.

I followed the documentation [here](https://www.metabase.com/docs/latest/users-guide/13-sql-parameters.html).
From admin panel I changed the columns’ type to Category.
Then I manually synced using database panel to make sure the changes take effect.
Basically what cmann50 had mentioned as fix.

When I went back to my question, `station_code` category widget was acting properly.
But, `delivery_mode` widget was not acting as Category (no drop down menu) rather a free text Input widget (whose interaction with field filters I do not know and it was resulting in errors anyway).

When can this happen? What can I try out further? I do not know whether columns allowing NULL or not should have an impact (have provided the info in schema). Even though delivery_mode allows NULL, no record has any NULL value at present.

To answer Sal's question above, none of the category fields have more than 10 distinct values.

**Comment by scottgrayson** [comment]
Is this the same issue or should I create a new issue? Here's my problem

# Category filter with redshift varchar causes [operator does not exist: character varying = boolean]

- Your browser and the version: Chrome 64
- Your operating system: OS X 10
- Your databases: Redshift
- Metabase version: 28.1
- Metabase hosting environment: Elastic Beanstalk, Docker
- Metabase internal database: Postgres

redshift table being queried
```sql
CREATE TABLE commercials
(
  id integer identity(0,1) encode raw,
  name VARCHAR(250) encode zstd,
  duration integer encode delta32k,
  first_seen timestamp encode delta32k,
  last_seen timestamp encode delta32k
)
-- distkey(id)
sortkey(id);
```

example metabase query
```
select count(*)
from commercials
where true
[[and name = {{ name }}]]
[[and duration = {{ duration }}]]
```
the duration filter works (if I comment out the name filter)

The error I get is
`[Amazon](500310) Invalid operation: operator does not exist: character varying = boolean`

**Comment by scottgrayson** [comment]
this change worked for me
```sql
[[and name = {{ name }}]]
```
changed to 
```sql
[[and {{name}}]]
```

full working query
```sql
select count(*)
from commercials
where true
[[and name = {{ name }}]]
[[and duration = {{ duration }}]]
```

**Comment by flamber** [comment]
I cannot reproduce on 0.35.4 - this was likely fixed in 0.29
